pkg/apis/apps/v1alpha1/uniteddeployment_types.go

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2019 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// UnitedDeploymentSpec defines the desired state of UnitedDeployment
+type UnitedDeploymentSpec struct {
+	// Replicas is the totally desired number of replicas of all the owning workloads.
+	// If unspecified, defaults to 0.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// selector is a label query over pods that should match the replica count.
+	// It must match the pod template's labels.
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// Template is the object that describes the subset that will be created
+	Template SubsetTemplate `json:"template,omitempty"`
+
+	// Contains the information of subset topology
+	Topology Topology `json:"topology,omitempty"`
+
+	Strategy UnitedDeploymentUpdateStrategy `json:"strategy,omitempty"`
+
+	// Indicate the number of histories to be conserved
+	// If unspecified, defaults to 20
+	// +optional
+	RevisionHistoryLimit *int32 `json:"historyLimit,omitempty"`
+}
+
+// SubsetTemplate defines the subset under the UnitedDeployment
+type SubsetTemplate struct {
+	// Indicate the type of the underlying workload
+	// If unspecified, defaults to ReplicaSet
+	// +optional
+	SubsetType *metav1.TypeMeta `json:"subsetType,omitempty"`
+
+	// StatefulSet template
+	StatefulSet *StatefulSetTemplateSpec `json:"statefulSet,omitempty"`
+
+	// ReplicaSet template
+	ReplicaSet *ReplicaSetTemplateSpec `json:"replicaSet,omitempty"`
+}
+
+// StatefulSetTemplateSpec defines the subset template of StatefulSet
+type StatefulSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.StatefulSetSpec `json:"spec,omitempty"`
+}
+
+// ReplicaSetTemplateSpec defines the subset template of ReplicasSet
+type ReplicaSetTemplateSpec struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              appsv1.ReplicaSetSpec `json:"spec,omitempty"`
+}
+
+// UnitedDeploymentUpdateStrategy defines the update strategy of UnitedDeployment
+type UnitedDeploymentUpdateStrategy struct {
+	// Pause the rollout of all subset worloads
+	// +optional
+	Pause bool `json:"pause,omitempty"`
+
+	// Indicate the partition of each subset
+	Partitions map[string]int32 `json:"partitions,omitempty"`
```

Comment:
Did you validate the key of the map matches a subset name? I didn't find it in uniteddeployment_validation.go (TODO in future change is fine).

Also please add a comment to explain this field is not used if template is replicaSet.

Reply:
> Also please add a comment to explain this field is not used if template is replicaSet.

How about we use the format `<deployment-name>-<subset-name>-` as the prefix of replicasSet name